File: backend/app/schemas/purchase.py
"""采购管理 Pydantic 模型。"""
from datetime import date
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict


class PurchaseItemBase(BaseModel):
    item_name: str
    spec: Optional[str] = None
    quantity: int = 1
    unit_price: Optional[Decimal] = None
    amount: Optional[Decimal] = None
    supplier: Optional[str] = None
    remark: Optional[str] = None


class PurchaseItemCreate(PurchaseItemBase):
    pass


class PurchaseItemRead(PurchaseItemBase):
    id: int
    req_id: int
    model_config = ConfigDict(from_attributes=True)


class PurchaseReqBase(BaseModel):
    req_no: Optional[str] = None
    applicant: str
    department: Optional[str] = None
    item_name: str
    spec: Optional[str] = None
    quantity: int = 1
    expected_amount: Optional[Decimal] = None
    supplier: Optional[str] = None
    expected_date: Optional[date] = None
    reason: Optional[str] = None
    status: str = "草稿"
    submit_date: Optional[date] = None
    approver: Optional[str] = None
    approve_date: Optional[date] = None
    approve_remark: Optional[str] = None
    is_rd_project: Optional[str] = None  # 是否归属研发项目：是/否
    rd_project_code: Optional[str] = None  # 研发项目编码
    remark: Optional[str] = None
    items: list[PurchaseItemCreate] = []


class PurchaseReqCreate(PurchaseReqBase):
    pass


class PurchaseReqUpdate(BaseModel):
    req_no: Optional[str] = None
    applicant: Optional[str] = None
    department: Optional[str] = None
    item_name: Optional[str] = None
    spec: Optional[str] = None
    quantity: Optional[int] = None
    expected_amount: Optional[Decimal] = None
    supplier: Optional[str] = None
    expected_date: Optional[date] = None
    reason: Optional[str] = None
    status: Optional[str] = None
    submit_date: Optional[date] = None
    approver: Optional[str] = None
    approve_date: Optional[date] = None
    approve_remark: Optional[str] = None
    is_rd_project: Optional[str] = None
    rd_project_code: Optional[str] = None
    remark: Optional[str] = None
    items: Optional[list[PurchaseItemCreate]] = None


class PurchaseReqRead(PurchaseReqBase):
    id: int
    items: list[PurchaseItemRead] = []
    model_config = ConfigDict(from_attributes=True)


class ApprovalBody(BaseModel):
    approver: str
    remark: Optional[str] = None

File: frontend/src/types/purchase.ts
export type ReqStatus = '草稿' | '待审批' | '已通过' | '已驳回'

/** 采购申请明细：一条申请可包含多个物品 / 服务 */
export interface PurchaseItem {
  id?: number
  req_id?: number
  item_name: string
  spec?: string | null
  quantity: number
  unit_price?: number | null
  amount?: number | null
  supplier?: string | null
  remark?: string | null
}

export interface PurchaseReq {
  id: number
  req_no?: string | null
  applicant: string
  department?: string | null
  item_name: string
  spec?: string | null
  quantity: number
  expected_amount?: number | null
  supplier?: string | null
  expected_date?: string | null
  reason?: string | null
  status: ReqStatus
  submit_date?: string | null
  approver?: string | null
  approve_date?: string | null
  approve_remark?: string | null
  is_rd_project?: string | null
  rd_project_code?: string | null
  remark?: string | null
  items?: PurchaseItem[]
}
